src/types/complex.rs

```diff
@@ -19,118 +19,224 @@ pyobject_native_type!(
+    fn abs(&self) -> c_double;
+    /// Returns `self` raised to the power of `other`.
+    #[cfg(not(any(Py_LIMITED_API, PyPy)))]
+    fn pow(&self, other: &Bound<'py, PyComplex>) -> Bound<'py, PyComplex>;
+}
+
+impl<'py> PyComplexMethods<'py> for Bound<'py, PyComplex> {
+    fn real(&self) -> c_double {
+        unsafe { ffi::PyComplex_RealAsDouble(self.as_ptr()) }
+    }
+
+    fn imag(&self) -> c_double {
         unsafe { ffi::PyComplex_ImagAsDouble(self.as_ptr()) }
     }
+
+    #[cfg(not(any(Py_LIMITED_API, PyPy)))]
+    fn abs(&self) -> c_double {
+        unsafe {
+            let val = (*(self.as_ptr() as *mut ffi::PyComplexObject)).cval;
+            ffi::_Py_c_abs(val)
+        }
+    }
+
+    #[cfg(not(any(Py_LIMITED_API, PyPy)))]
+    fn pow(&self, other: &Bound<'py, PyComplex>) -> Bound<'py, PyComplex> {
+        use crate::ffi_ptr_ext::FfiPtrExt;
+        unsafe {
+            not_limited_impls::complex_operation(
+                self.as_borrowed(),
+                other.as_borrowed(),
+                ffi::_Py_c_pow,
+            )
+            .assume_owned(self.py())
+            .downcast_into_unchecked()
+        }
+    }
 }
 
 #[cfg(not(any(Py_LIMITED_API, PyPy)))]
 mod not_limited_impls {
+    use crate::ffi_ptr_ext::FfiPtrExt;
+    use crate::Borrowed;
+
     use super::*;
     use std::ops::{Add, Div, Mul, Neg, Sub};
 
     impl PyComplex {
         /// Returns `|self|`.
         pub fn abs(&self) -> c_double {
-            unsafe {
-                let val = (*(self.as_ptr() as *mut ffi::PyComplexObject)).cval;
-                ffi::_Py_c_abs(val)
-            }
+            self.as_borrowed().abs()
         }
         /// Returns `self` raised to the power of `other`.
-        pub fn pow(&self, other: &PyComplex) -> &PyComplex {
-            unsafe {
-                self.py()
-                    .from_owned_ptr(complex_operation(self, other, ffi::_Py_c_pow))
-            }
+        pub fn pow<'py>(&'py self, other: &'py PyComplex) -> &'py PyComplex {
+            self.as_borrowed().pow(&other.as_borrowed()).into_gil_ref()
         }
     }
 
     #[inline(always)]
-    unsafe fn complex_operation(
-        l: &PyComplex,
-        r: &PyComplex,
+    pub(super) unsafe fn complex_operation<'py>(
+        l: Borrowed<'_, 'py, PyComplex>,
+        r: Borrowed<'_, 'py, PyComplex>,
         operation: unsafe extern "C" fn(ffi::Py_complex, ffi::Py_complex) -> ffi::Py_complex,
     ) -> *mut ffi::PyObject {
         let l_val = (*(l.as_ptr() as *mut ffi::PyComplexObject)).cval;
         let r_val = (*(r.as_ptr() as *mut ffi::PyComplexObject)).cval;
         ffi::PyComplex_FromCComplex(operation(l_val, r_val))
     }
 
-    impl<'py> Add for &'py PyComplex {
-        type Output = &'py PyComplex;
-        fn add(self, other: &'py PyComplex) -> &'py PyComplex {
-            unsafe {
-                self.py()
-                    .from_owned_ptr(complex_operation(self, other, ffi::_Py_c_sum))
+    macro_rules! bin_ops {
```

Comment:
Nice use of a macro here 👍 

(There are sooo many implementations of these traits for the various combinations of borrowed forms 😂)

Reply:
Yeah, i started with handwritten impls, but that got out of hand quickly 🙃